### PR TITLE
[SCHEMA] Add metadata term files

### PR DESCRIPTION
**Pull request by tsalo**
References #699.

This is a draft that might be useful for determining how to structure metadata YAML files for the schema. 

This approach does not incorporate any interactions between metadata terms, or requirement levels, which are both dependent on the suffix or data type to which the metadata term is being applied. This fits well with #603.

Changes proposed:

- Take the metadata JSON files from [the bids-validator](https://github.com/bids-standard/bids-validator/tree/master/bids-validator/validators/json/schemas) and convert them to YAML format.
- Add "description" and "key_name" fields to the metadata YAML files.
- Add functions to render the metadata information in tables within the schema. These functions require the specification text to include the requirement levels.
- Create files for a number of metadata terms that aren't present in the validator, but _are_ present in specification tables.

TODO:

- [x] Different units dependent on different data types?
    - Yes
- [x] Include "unit" for string/integer data types?
    - No
- [x] Limits, like minimum and maximum values
    - Taken from validator
- [x] Should we retain references to other info (e.g., lists of valid coordinate spaces for iEEG, EEG, and MEG)?
    - Currently, I have the references still in the files, with the referenced info in "private" yaml files (i.e., files with leading underscore).
- [x] Links that have to be broken up aren't looking great.
    - This is because the flattening step adds a space between separate lines.
- [x] Would be great to replace relative link paths with absolute ones.
- [X] Type determination for objects.
    - `additionalProperties` defines type.
    - `properties` should create additional rows. How should individual sub-object-level requirement levels be annotated then?? Take `Genetics.Dataset`, `Genetics.Database`, and `Genetics.Descriptors` as an example.
        - SOLUTION: Instead of trying to expand nested dictionaries into new rows, I've replaced the internal definitions with `$ref`s and added new files for the properties that can be referenced on their own. So now Genetics' `properties` fields is a dict with three `$ref`s and there are separate files for `Genetics.Dataset.yaml` etc.
- [x] Make type determination function recursive to deal with nested objects
    - See `SoftwareFilters` as an example.
- [x] Document new structure once #740 is merged.
- [x] Add "relative" and "absolute"(?) path formats for strings?
    - I decided to go with `dataset_relative` and `participant_relative` for now.
- [x] Add "unit" format for strings?
    - With PET now merged in, we have a bunch of new metadata fields that just define units of other fields instead of committing to a standard like we have for most other fields with units. Maybe we could define a "unit" format that will let the validator check for valid units?
- [ ] Identify and incorporate section-specific definition modifications.

### CODE REVIEW

File: src/schema/metadata/PostLabelingDelay.yaml
Comment (tsalo) [inline src/schema/metadata/PostLabelingDelay.yaml:16]:
```suggestion
  - number:
    - minimum: 0
```

What about this kind of notation for limits?

File: src/schema/metadata/PostLabelingDelay.yaml
Comment (tsalo) [inline src/schema/metadata/PostLabelingDelay.yaml:17]:
```suggestion
  - array:
      number:
        minimum: 0
```

Maybe this for nesting rules for the data types within the array?

**Comment by effigies** [comment]
Is there an advantage to a single file per term? Grouping terms feels more natural to me.

**Comment by tsalo** [comment]
I don't want to group terms into multiple files, since relationships between terms (including their categories) probably vary across the specification, along with any logic that needs to be applied. I figured one file per term would be easier to search and edit than one file for all terms.

**Comment by tsalo** [comment]
Based on the current code, the following macro call in the specification produces the table at the bottom. Some are missing descriptions right now and others are just missing completely, but it's still pretty cool!

```Text
{{ MACROS___make_metadata_table(
   {
      "Manufacturer": "RECOMMENDED",
      "ManufacturersModelName": "RECOMMENDED",
      "DeviceSerialNumber": "RECOMMENDED",
      "StationName": "RECOMMENDED",
      "SoftwareVersions": "RECOMMENDED",
      "HardcopyDeviceSoftwareVersion": "DEPRECATED",
      "MagneticFieldStrength": "RECOMMENDED, but REQUIRED for Arterial Spin Labeling",
      "ReceiveCoilName": "RECOMMENDED",
      "ReceiveCoilActiveElements": "RECOMMENDED",
      "GradientSetType": "RECOMMENDED",
      "MRTransmitCoilSequence": "RECOMMENDED",
      "MatrixCoilMode": "RECOMMENDED",
      "CoilCombinationMethod": "RECOMMENDED",
   }
) }}
```

image

File: src/04-modality-specific-files/01-magnetic-resonance-imaging-data.md
Comment (tsalo) [inline src/04-modality-specific-files/01-magnetic-resonance-imaging-data.md:29]: Until we add conditional logic to the schema, this would be how we define requirement levels in the specification.

File: src/schema/metadata/AnatomicalLandmarkCoordinateSystem.yaml
Comment (tsalo) [inline src/schema/metadata/AnatomicalLandmarkCoordinateSystem.yaml:13]: I'm not so sure about this one. Do we want to include references in the schema? If so, is this the best way to do it?

**Comment by tsalo** [comment]
I can keep trucking along, adding terms, but I'd like to get input from interested folks.

@satra @dbkeator would this format be useable for NIDM-Terms?

@yarikoptic I know this isn't exactly how you wanted it (e.g., files are grouped in subfolders by type), but WDYT? The actual interactions between terms will still need to be defined elsewhere in the schema.

@rwblair could the validator use files like this? They're basically a translation of the validator JSON files to individual YAML files, with added descriptions and units, so I assume it wouldn't require too much effort to support them.

File: src/schema/metadata/AnatomicalLandmarkCoordinateSystem.yaml
Comment (tsalo) [reply src/schema/metadata/AnatomicalLandmarkCoordinateSystem.yaml]: Alternatively, we could make sure that all referenced definitions are in "private" files (e.g., `_ StandardTemplateCoordSys.yaml`), since those definitions shouldn't be useful on their own.

File: src/02-common-principles.md
Comment (yarikoptic) [inline src/02-common-principles.md:510]: table seems to be (rendered) empty, shortcut link https://bids-specification--762.org.readthedocs.build/en/762/02-common-principles.html#tabular-files

File: src/03-modality-agnostic-files.md
Comment (yarikoptic) [inline src/03-modality-agnostic-files.md:19]: I don't see name in rendered version: https://bids-specification--762.org.readthedocs.build/en/762/03-modality-agnostic-files.html#dataset_descriptionjson

File: src/03-modality-agnostic-files.md
Comment (yarikoptic) [inline src/03-modality-agnostic-files.md:92]: SourceDatasets is missing: https://bids-specification--762.org.readthedocs.build/en/762/03-modality-agnostic-files.html#derived-dataset-and-pipeline-description

I guess script should be adjusted to error out if some field is not "handled"?

File: src/03-modality-agnostic-files.md
Comment (yarikoptic) [inline src/03-modality-agnostic-files.md:31]: having such specifications in the document text would be a big step back from toward the target of having a "machine readable" schema.

Why not to add `metadata` key to the corresponding records, e.g. in this case of https://github.com/bids-standard/bids-specification/blob/master/src/schema/top_level_files.yaml#L14 where similar to `entities` of https://github.com/bids-standard/bids-specification/tree/master/src/schema/datatypes to list the 'requirements' for each one of those? and then in the macros just point to what metadata of to render here (e.g. of `top_level_files.dataset_description` here)?  or you can immediately see that this would not work?

File: src/04-modality-specific-files/04-intracranial-electroencephalography.md
Comment (yarikoptic) [inline src/04-modality-specific-files/04-intracranial-electroencephalography.md:442]: oh, that is an interesting case...  in light of my eager desire for concentrating this information in machine readable form, I can suggest for such cases to get to temp middle point  where requirement could be "spelled out" in human form for now with the last one listed as "catch all", e.g. 

```yaml
metadata:
 iEEGCoordinateUnits: required
 iEEGCoordinateUnits: required
 iEEGCoordinateSystemDescription:
  required: if `iEEGCoordinateSystem` is `"Other"`
  recommended:
 iEEGCoordinateProcessingDescription: recommended
```

**Review by yarikoptic — CHANGES_REQUESTED** [review]
> @yarikoptic I know this isn't exactly how you wanted it (e.g., files are grouped in subfolders by type), but WDYT? The actual interactions between terms will still need to be defined elsewhere in the schema.

GREAT STUFF HERE, thank you @tsalo for all the work on this!  I blissfully forgotten (and decided to not encumber my review with prior opinion by looking it up) what I wanted.  But I immediately saw what I would have really preferred to avoid here -- "leaking" schema into the document.   I have dropped some question/idea(s) in other comments on how possibly may be those metadata fields "requirements" could be centralized within schema.  WDYT?

File: src/03-modality-agnostic-files.md
Comment (tsalo) [reply src/03-modality-agnostic-files.md:92]: Ultimately yes, but since I'm still in the process of converting all of the metadata terms, I don't want everything to error out just yet. Right now it just filters out any missing terms and renders the available ones.

File: src/03-modality-agnostic-files.md
Comment (tsalo) [reply src/03-modality-agnostic-files.md:31]: > having such specifications in the document text would be a big step back from toward the target of having a "machine readable" schema.

I see it as an incremental move forward. It definitely doesn't solve a lot of the problems (e.g., requirement levels, conditional relationships between metadata fields), but to be fair those are the _hard_ problems so I don't feel too bad about it.

> Why not to add metadata key to the corresponding records, e.g. in this case of https://github.com/bids-standard/bids-specification/blob/master/src/schema/top_level_files.yaml#L14 where similar to entities of https://github.com/bids-standard/bids-specification/tree/master/src/schema/datatypes to list the 'requirements' for each one of those? and then in the macros just point to what metadata of to render here (e.g. of top_level_files.dataset_description here)? or you can immediately see that this would not work?

I'll have to try it out, but that sounds like a good plan for the PR _after_ this one. I figure that, since we want to separate the core definitions from the rules of the schema, we can tackle those two pieces in separate stages.

Comment (tsalo) [reply src/03-modality-agnostic-files.md:31]: I think the best next step for the schema (after this) is to try to identify the types of rules we can expect from the schema, along with any really weird edge cases. That's what I've tried to do in #620, although I worry that my current list isn't exhaustive. Without a list of necessary rules, I'm not sure if we can define them in YAML files in a way that will satisfy everyone.

Comment (yarikoptic) [reply src/03-modality-agnostic-files.md:31]: Sure could and should be a multistage process. I am ok with your proposal. I just thought that  it might be nice to just put rules identification for later pr, thus keep description free form, while already allowing for machine readable specification of possible requirement levels and keeping everything under src/schema. But such a move could indeed be just yet another follow up PR to not block this one

**Comment by rwblair** [comment]
> @rwblair could the validator use files like this? They're basically a translation of the validator JSON files to individual YAML files, with added descriptions and units, so I assume it wouldn't require too much effort to support them.

The validator certainly could build its json schema files from this. Is there any where encoded in yaml which terms are acceptable in which json files?

**Comment by tsalo** [comment]
>Is there any where encoded in yaml which terms are acceptable in which json files?

Not yet. I think I'll need to translate over the context files from the validator, with some minor additions.

**Comment by sappelhoff** [comment]
I think the URL the linkchecker flags has moved to `https://www.hedtags.org/docs/HEDTaggingStrategyGuide.pdf`, we should probably fix that in a PR that will be merged soon, or make a dedicated PR

**Comment by tsalo** [comment]
>The validator certainly could build its json schema files from this. Is there any where encoded in yaml which terms are acceptable in which json files?

@rwblair, I'm sorry, it looks like I misunderstood your question. Right now the _rules_ applied to the metadata terms (e.g., which terms apply to each suffix or datatype) are not implemented, and I wasn't planning to put that in this PR. I figured that the validator could retain those rules for now, but without the information that is in the YAML files I'm proposing.

I'm hopeful that someone will be able to tackle the rules (and thus #699) in a future PR.

**Comment by tsalo** [comment]
Okay, I think I'm happy with this, so I'm marking it as ready for review.

File: src/schema/metadata/PowerLineFrequency.yaml
Comment (tsalo) [inline src/schema/metadata/PowerLineFrequency.yaml:11]: Could we replace the `^n/a$` pattern with a one-item `enum`?

File: src/schema/metadata/BackgroundSuppressionNumberPulses.yaml
Comment (tsalo) [inline src/schema/metadata/BackgroundSuppressionNumberPulses.yaml:8]: Is `minimum` equivalent to `inclusiveMinimum`?

File: src/schema/metadata/BolusCutOffDelayTime.yaml
Comment (tsalo) [inline src/schema/metadata/BolusCutOffDelayTime.yaml:14]: I've noticed there's both `minimum` and `Minimum`.

File: src/schema/metadata/IntendedFor.yaml
Comment (tsalo) [inline src/schema/metadata/IntendedFor.yaml:10]: What if we added a `format: relative_path` option?

File: src/schema/metadata/MEGCoordinateSystem.yaml
Comment (tsalo) [inline src/schema/metadata/MEGCoordinateSystem.yaml:13]: This currently renders as "string or string or string or string". Not sure how best to clean it up.

File: src/schema/metadata/Genetics.yaml
Comment (tsalo) [inline src/schema/metadata/Genetics.yaml:12]: What does everyone think of this instead of defining those fields in here directly?

File: src/schema/metadata/MEGCoordinateSystem.yaml
Comment (rwblair) [reply src/schema/metadata/MEGCoordinateSystem.yaml:13]: I tried rendering it with the enum typed fields joined with " or ", unfortunately the resulting list seems unwieldy to put directly into the table. This also wouldn't make sense for other fields like GeneticLevel where the enumerations are listed directly in the description.

Is there any current situation where removing duplicates of  what `_resolve_metadata_type` returns when generating the string for anyOf would cause a problem?

Comment (tsalo) [reply src/schema/metadata/MEGCoordinateSystem.yaml:13]: Currently it's just a single single that's appended to, so it's difficult to identify and remove duplicates. I could reorganize it to a list of lists or something that would be easier to identify and remove duplicates though.

**Review by sappelhoff — COMMENTED** [review]
How can we most effectively review these 234 files? Do you have a "guide" @tsalo ? Where is another set of eyes most useful?

File: src/schema/metadata/PowerLineFrequency.yaml
Comment (sappelhoff) [reply src/schema/metadata/PowerLineFrequency.yaml:11]: yes, technically that would be equivalent except for the message that a user gets upon validating that schema. (I.e., "foo must be pattern X" versus "foo must be one of X" or the like). But that small difference would be no problem IMHO.

**Comment by tsalo** [comment]
I think the best place to start is the inline comments/questions I added. I also added a lot of information to `src/schema/README.md` that I _hope_ is accurate, and I'd love input on it. Other than that, I guess the best things to do would be to check the generated tables, which are rendered alongside the original tables for easy comparison.

File: src/schema/metadata/PowerLineFrequency.yaml
Comment (tsalo) [reply src/schema/metadata/PowerLineFrequency.yaml:11]: The current issue is that the tables render these kinds as `object or string` instead of `object or "n/a"`. I thought maybe checking against an enum list and rendering those values directly if the list is short would be easier than figuring something out for pattern strings, but maybe it'll be easier if I figure out what the characters in the pattern represent, and what patterns are possible?

File: src/schema/metadata/IntendedFor.yaml
Comment (tsalo) [reply src/schema/metadata/IntendedFor.yaml:10]: I've added `dataset_relative`, `participant_relative`, and `unit` formats. They may not be used yet, but they could ultimately be useful, I think.

File: src/schema/metadata/Columns.yaml
Comment (tsalo) [inline src/schema/metadata/Columns.yaml:8]: Could we just assume that strings have a minimum length of 1? If the field is provided with a string, then I think it should have at least one character to be valid, from a BIDS standpoint. Missing data should be provided with `"n/a"`, so there's not really a case for `""` appearing in jsons.

File: src/schema/metadata/DatasetDOI.yaml
Comment (tsalo) [inline src/schema/metadata/DatasetDOI.yaml:8]: I started replacing internal relative links (e.g., `../02-common-principles.html#definitions`) with absolute versions (`/02-common-principles.html#definitions`). How does everyone feel about this change? I think it would be good to apply throughout the whole specification.

File: src/04-modality-specific-files/09-positron-emission-tomography.md
Comment (tsalo) [inline src/04-modality-specific-files/09-positron-emission-tomography.md:152]: I _think_ this was a typo in the PET BEP?

File: src/04-modality-specific-files/03-electroencephalography.md
Comment (tsalo) [inline src/04-modality-specific-files/03-electroencephalography.md:414]: What does everyone think of this approach for extending descriptions for modality- or table-specific purposes?

File: src/schema/metadata/PowerLineFrequency.yaml
Comment (sappelhoff) [reply src/schema/metadata/PowerLineFrequency.yaml:11]: > maybe it'll be easier if I figure out what the characters in the pattern represent, and what patterns are possible?

There are some libraries that allow you to generate strings from a given regular expression, see: https://github.com/asciimoo/exrex

So perhaps it wouldn't even be too difficult. And coincidentally, it may provide us with a nice double check of our regexps :-)

File: src/schema/metadata/BackgroundSuppressionNumberPulses.yaml
Comment (sappelhoff) [reply src/schema/metadata/BackgroundSuppressionNumberPulses.yaml:8]: > Is minimum equivalent to inclusiveMinimum?

conceptually yes, but according to the JSON schema spec, there is no dedicated keyword for `inclusiveMinimum`: https://json-schema.org/understanding-json-schema/reference/numeric.html?highlight=minimum#range



        x ≥ minimum
        x > exclusiveMinimum
        x ≤ maximum
        x < exclusiveMaximum

Comment (sappelhoff) [reply src/schema/metadata/BackgroundSuppressionNumberPulses.yaml:8]: related: https://github.com/bids-standard/bids-validator/pull/1261

File: src/schema/metadata/BolusCutOffDelayTime.yaml
Comment (sappelhoff) [reply src/schema/metadata/BolusCutOffDelayTime.yaml:14]: That's weird, I am pretty sure that in JSON schema, the keys are case sensitive, and that `minimum` is a key and `Minimum` is not. I am just surprised that we did not run into errors so far.

will fix in https://github.com/bids-standard/bids-validator/pull/1261

File: src/schema/metadata/Genetics.yaml
Comment (sappelhoff) [reply src/schema/metadata/Genetics.yaml:12]: looks clean to me

File: src/schema/metadata/Columns.yaml
Comment (sappelhoff) [reply src/schema/metadata/Columns.yaml:8]: agreed! And then users can still abuse it by providing `" "`, but it might be sufficient to make them think twice :-)

File: src/schema/metadata/DatasetDOI.yaml
Comment (sappelhoff) [reply src/schema/metadata/DatasetDOI.yaml:8]: Why are you doing that? I thought it's an advantage to point to the relative files and have Mkdocs generate the suitable links for us? Pointing to html directly feels error prone but I can't put my finger on it right now :thinking:

File: src/04-modality-specific-files/09-positron-emission-tomography.md
Comment (sappelhoff) [reply src/04-modality-specific-files/09-positron-emission-tomography.md:152]: yes, most likely. fix: https://github.com/bids-standard/bids-specification/pull/771

File: src/schema/metadata/Columns.yaml
Comment (tsalo) [reply src/schema/metadata/Columns.yaml:8]: I think the same can apply to `minItems` when the type is `array`, so I'll remove all cases of `minLength: 1` and `minItems: 1` from the YAML files.

File: src/schema/metadata/BackgroundSuppressionNumberPulses.yaml
Comment (tsalo) [reply src/schema/metadata/BackgroundSuppressionNumberPulses.yaml:8]: Awesome! I'll replace `Minimum` with `minimum` and `inclusiveMaximum` with `maximum`. Thanks!

File: src/schema/metadata/DatasetDOI.yaml
Comment (tsalo) [reply src/schema/metadata/DatasetDOI.yaml:8]: One reason is that relative links depend on where the metadata field is rendered, so what might work for one file might not work for another. It should also make it easier to resolve links in external applications (e.g., NIDM-Terms).

Comment (sappelhoff) [reply src/schema/metadata/DatasetDOI.yaml:8]: I see :thinking: I couldn't find info on whether that's good or bad: https://www.mkdocs.org/user-guide/writing-your-docs/#internal-links

it's not really mentioned in the docs over there. So maybe we should carefully try it, and if it doesn't break anything it's fine.

**Pull request merged**
